bin/faucet/src/client.rs

```diff
@@ -0,0 +1,382 @@
+            account_seed,
+            root_block_header,
+            root_chain_mmr,
+        );
+        let authenticator = FaucetAuthenticator::new(secret);
+        let mut executor =
+            TransactionExecutor::new(data_store.clone(), Some(Rc::new(authenticator)));
+
+        executor
+            .load_account(id)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        let mut rng = thread_rng();
+        let coin_seed: [u64; 4] = rng.gen();
+        let rng = RpoRandomCoin::new(coin_seed.map(Felt::new));
+        Ok(Self { data_store, rpc_api, executor, id, rng })
+    }
+
+    pub fn execute_mint_transaction(
+        &mut self,
+        target_account_id: AccountId,
+        is_private_note: bool,
+        asset_amount: u64,
+    ) -> Result<(ExecutedTransaction, Note), FaucetError> {
+        let asset = FungibleAsset::new(self.id, asset_amount)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        let note_type = if is_private_note {
+            NoteType::OffChain
+        } else {
+            NoteType::Public
+        };
+
+        let output_note =
+            create_p2id_note(self.id, target_account_id, vec![asset.into()], note_type, self.rng)
+                .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        let transaction_args =
+            build_transaction_arguments(&output_note, &self.executor, note_type, asset)?;
+
+        let executed_tx = self
+            .executor
+            .execute_transaction(self.id, 0, &[], transaction_args)
+            .map_err(|err| {
+                FaucetError::InternalServerError(format!("Failed to execute transaction: {}", err))
+            })?;
+
+        Ok((executed_tx, output_note))
+    }
+
+    pub async fn prove_and_submit_transaction(
+        &mut self,
+        executed_tx: ExecutedTransaction,
+    ) -> Result<(), FaucetError> {
+        let transaction_prover = TransactionProver::new(ProvingOptions::default());
+
+        let delta = executed_tx.account_delta().clone();
+
+        let proven_transaction =
+            transaction_prover.prove_transaction(executed_tx).map_err(|err| {
+                FaucetError::InternalServerError(format!("Failed to prove transaction: {}", err))
+            })?;
+
+        let request = SubmitProvenTransactionRequest {
+            transaction: proven_transaction.to_bytes(),
+        };
+
+        self.rpc_api
+            .submit_proven_transaction(request)
+            .await
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        self.data_store.update_faucet_account(&delta).map_err(|err| {
+            FaucetError::InternalServerError(format!("Failed to update account: {}", err))
+        })?;
+
+        Ok(())
+    }
+
+    pub fn get_faucet_id(&self) -> AccountId {
+        self.id
+    }
+}
+
+#[derive(Clone)]
+pub struct FaucetDataStore {
+    faucet_account: Rc<RefCell<Account>>,
```

Comment:
If we changed `Rc<RefCell<Account>>` for `Arc<<Mutex<Account>>` (maybe even `Arc<RwLock<Account>>`) would we still need the `unsafe impl Send` and `unsafe impl Sync` from above?

Reply:
Yes because the `TransactionExecutor` still wouldn't be `Send+Sync`